### PR TITLE
12  get full day menu

### PR DESCRIPTION
**Pull request by kosyfrances**
#12 

* Add more commands to help
* Add `show timetable` command to list available timetables
* Add the following commands to retrieve meals.
```
GET MENU: menu
(if available timetable is just one, you will get the menu for today.
Else you will be given a list of available timetables to get choose from.)

GET MENU FOR A TIMETABLE: menu TIMETABLE_NAME
(For example: menu lagos
By default this gives you menu for today for the specific timetable given)

GET MENU FOR A TIMETABLE TODAY: menu TIMETABLE_NAME today

GET MENU FOR YESTERDAY FOR A TIMETABLE: menu TIMETABLE_NAME yesterday

GET MENU FOR TOMORROW FOR A TIMETABLE: menu TIMETABLE_NAME tomorrow

GET MENU FOR THE ANY DAY FOR A TIMETABLE: menu TIMETABLE_NAME DAY_OF_WEEK
(For example: menu TIMETABLE_NAME monday)
```
* Write tests

### CODE REVIEW

File: common/utils.py [hunk 1]
```diff
@@ -52,7 +60,36 @@ def get_date(day_arg):
...
+    api_url = dotenv.get('API_URL_ENDPOINT')
+    endpoint = '{}?query={}'.format(api_url, query)
+    headers = {'X-TavernaToken': dotenv.get('X-TAVERNATOKEN')}
+    return requests.post(endpoint, headers=headers).json()
+
+
+def make_api_request_for_timetables():
+    query = 'query{timetables{edges{node{name, slug, cycleLength,refCycleDay, \
+             vendors{edges{node{name}}}, admins{edges{node{username}}}}}}}'
+    return make_api_request(query)['timetables']
```
Comment (nbyodai) [inline common/utils.py]: I haven't really done extensive testing but seemed to have broken on my end.
`make_api_request(query)` returns 
`{
    "timetables": {
        "edges": []
    }
}`
and therefore `make_api_request_for_timetables` returns a dict not a list that line 94 is expecting.

After changing to `make_api_request(query)['timetables']['edges']` it stopped breaking but still tests were passing. I expected the tests to fail.

Comment (nbyodai) [reply common/utils.py]: Also, it might be good to check for keys in dicts before trying to access them.

File: common/utils.py [hunk 2]
```diff
@@ -52,7 +60,36 @@ def get_date(day_arg):
...
+
+def make_api_request(query):
+    api_url = dotenv.get('API_URL_ENDPOINT')
+    endpoint = '{}?query={}'.format(api_url, query)
+    headers = {'X-TavernaToken': dotenv.get('X-TAVERNATOKEN')}
+    return requests.post(endpoint, headers=headers).json()
+
+
+def make_api_request_for_timetables():
+    query = 'query{timetables{edges{node{name, slug, cycleLength,refCycleDay, \
```
Comment (nbyodai) [inline common/utils.py]: It seems like our queries can be seen as constants. what do you thing about 
`TIMETABLE_QUERY` = `'query{timetables{edges{node{name, slug, cycleLength,refCycleDay,  vendors{edges{node{name}}}, admins{edges{node{username}}}}}}}'`

Might be a nice to have

File: common/utils.py (hunk 1 above)
Comment (nbyodai) [reply common/utils.py]: No need, because of the way our queries are structured in GraphQL. Please ignore

**Comment by kosyfrances** [comment]
@andela-cobiora please continue with review

File: slackbot_settings.py [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
+
 dotenv.load()
 
 API_TOKEN = dotenv.get('SLACKBOT_API_TOKEN')
-DEFAULT_REPLY = 'Sorry, I have not been fully configured yet.'
+
+DEFAULT_REPLY = render('help_response.j2')
 
 # TO DO:
 # The value for ERRORS_TO should be specified in the env file later on
```
Comment (amos-o) [inline slackbot_settings.py]: Can't this be changed now?

Comment (kosyfrances) [reply slackbot_settings.py]: We can change it anytime @andela-aomondi. We can also change it now.

File: common/utils.py [hunk 4]
```diff
@@ -1,58 +1,103 @@
...
+        context['error'] = error
     return jinja2.Environment(
         loader=jinja2.FileSystemLoader(path)
     ).get_template(filename).render(context)
+
+
+def make_api_request(query):
+    api_url = dotenv.get('API_URL_ENDPOINT')
+    endpoint = '{}?query={}'.format(api_url, query)
+    headers = {'X-TavernaToken': dotenv.get('X-TAVERNATOKEN')}
```
Comment (nbyodai) [inline common/utils.py:87]: so right now we are making the request using API token in `.env` file.
What is the process for users who don't have access to the slack bot?

File: plugins/menu_plugin.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+    def make_api_request_for_events(self):
+        query = 'query {events{edges{node{name, action, startDate, endDate}}}}'
+        res = make_api_request(query)
+        events = res.get('events')
+        if events and 'edges' not in events:
+            return events
+
+    def servings_to_dict(self, servings):
+        new_dict = {}
```
Comment (nbyodai) [inline plugins/menu_plugin.py]: can use a more expressive name here

File: plugins/menu_plugin.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+        for obj in servings:
+            meal = obj['menuItem']['meal']['name']
+            menu_item = obj['menuItem']
+            new_obj = {
```
Comment (nbyodai) [inline plugins/menu_plugin.py]: can use a more expressive variable name in place of `new_obj` here

File: plugins/menu_plugin.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                value,
+                key=itemgetter('sequence_order')
+            ) for key, value in new_dict.items()
+        }
+
+    def get_meals(self, timetable, day):
+        """
+        Get meals for a particular timetable.
+        day is a string either today, tomorrow, yesterday, or any weekday
+        """
```
Comment (nbyodai) [inline plugins/menu_plugin.py:61]: add new line after

File: plugins/menu_plugin.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            meals = self.servings_to_dict(servings)
+            return meals
+
+    def get_event(self, day):
+        events = self.make_api_request_for_events()
+        event_list = []
+
+        if events:
+            date = parser.parse(self.date_helper.date_to_str(day)).isoformat()
+            for event in events:
```
Comment (nbyodai) [inline plugins/menu_plugin.py]: could move `parser.parse` and `isoformat` method call to a separate function and use that here.

File: plugins/menu_plugin.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+                start_date = parser.parse(event['startDate']).isoformat()
+                end_date = parser.parse(event['endDate']).isoformat()
+
+                if (date >= start_date) and (date <= end_date):
```
Comment (nbyodai) [inline plugins/menu_plugin.py]: `if (start_date<=date<=end_date): ...`

File: plugins/menu_plugin.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    # Convert message text to list to remove multiple spaces that may have
+    # been mistakenly added by the user and convert the list back to string
+    message_text_list = message.body['text'].lower().split()
+    len_msg_text_list = len(message_text_list)
+    message_text = ' '.join(message_text_list)
+    timetable_res = TimetableAPIUtils()
+    timetable_names = timetable_res.list_timetable_names()
+
+    if len_msg_text_list > 1:
+        timetable_name = message_text_list[1]
```
Comment (nbyodai) [inline plugins/menu_plugin.py]: is it possible for the value of timetable_name to be undefined?

File: plugins/menu_plugin.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+    if len_msg_text_list > 2:
+        day_of_week = message_text_list[2]
+
+    num_timetables = len(timetable_names)
```
Comment (nbyodai) [inline plugins/menu_plugin.py]: can we move to just after line 102

File: plugins/menu_plugin.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        if num_timetables < 1:
+            context.update({'no_timetable': True})
+        elif num_timetables == 1:
+            meals = menu_helper.get_meals(timetable_names[0], 'today')
+            menu_helper.meals_check_context_update(meals, context, 'today')
+        else:
+            context.update({'multiple_timetables': True})
+
+        response = render('menu_response.j2', context)
+        message.reply(response)
```
Comment (nbyodai) [inline plugins/menu_plugin.py]: is `message.reply` like `return` in that it breaks control flow?
I'm thinking we could break up the logic in this long chain of `if...elif..else` statements

Comment (nbyodai) [reply plugins/menu_plugin.py]: and with a breakdown we can provide better error messages for the user and not the general help message

File: common/utils.py (hunk 4 above)
Comment (kosyfrances) [reply common/utils.py:87]: well, they would have to ask us for a token, or gain access to a slackbot and try to test.

File: plugins/menu_plugin.py (hunk 10 above)
Comment (kosyfrances) [reply plugins/menu_plugin.py]: Not in this case, hence the reason for checking the length of the list of message being sent in by the user. We would use it to check if that timetable given exists.

File: plugins/menu_plugin.py (hunk 12 above)
Comment (kosyfrances) [reply plugins/menu_plugin.py]: @andela-cobiora `message.reply` is like return.

**Comment by kosyfrances** [comment]
@andela-cobiora @amos-o changes updated. Please continue with review .... same with you @kola-er :)

**Review by kola-er — APPROVED** [review]
Super! Welldone @kosyfrances.

**Review by nbyodai — APPROVED** [review]

**Pull request merged**